File: XUCore/ui/views/XUPositionedWindowView.m
//
//  XUPositionedWindowView.m
//  Downie
//
//  Created by Charlie Monroe on 9/10/14.
//  Copyright (c) 2014 Charlie Monroe Software. All rights reserved.
//

#import "XUPositionedWindowView.h"

#import "XUAbstract.h"

#import "NSArrayAdditions.h"
#import "FCLocalizationSupport.h"

@interface NSView (Secret)
-(void)_addKnownSubview:(NSView*)view;
@end

@implementation XUPositionedWindowView

-(void)_updateFramePosition{
	[self setFrame:[self frame]];
	
	[self setAutoresizingMask:NSViewMinXMargin];
}
-(void)awakeFromNib{
	if ([self shouldBeConnectedOnAwakeFromNib]){
		NSView *superview = [[_connectedToWindow contentView] superview];
		if ([_connectedToWindow respondsToSelector:@selector(addTitlebarAccessoryViewController:)]){
			if ([[_connectedToWindow titlebarAccessoryViewControllers] count] == 0){
				/** Since 10.10.2, the titlebar view is deferred in creation.
				 * We need to force the creation by adding a blank titlebar accessory
				 * view controller.
				 */
				NSTitlebarAccessoryViewController *controller = [[NSTitlebarAccessoryViewController alloc] init];
				[controller setLayoutAttribute:NSLayoutAttributeRight];
				[controller setView:[[NSView alloc] init]];
				[_connectedToWindow addTitlebarAccessoryViewController:controller];
			}
			
			superview = [[superview subviews] find:^BOOL(NSView *view) {
				return [view isKindOfClass:NSClassFromString(@"NSTitlebarContainerView")];
			}];
			
			superview = [[superview subviews] find:^BOOL(NSView *view) {
				return [view isKindOfClass:NSClassFromString(@"NSTitlebarView")];
			}];
			
			[superview addSubview:self];
		}else{
			[superview addSubview:self];
		}
	}
	
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_updateFramePosition) name:NSWindowDidResizeNotification object:_connectedToWindow];
	[self _updateFramePosition];
	
	[self localizeView];
}
-(void)dealloc{
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}
-(NSRect)frame{
	return [self frameForWindowBounds:[[self superview] bounds] andRealFrame:[super frame]];
}
-(NSLayoutAttribute)layoutAttribute{
	return NSLayoutAttributeBottom;
}
-(BOOL)postsFrameChangedNotifications{
	return NO;
}
-(BOOL)shouldBeConnectedOnAwakeFromNib{
	return YES;
}

XUGenerateAbstractMethod(-(NSRect)frameForWindowBounds:(NSRect)bounds andRealFrame:(NSRect)realFrame);

@end

File: XUCore/ui/views/XUPositionedWindowView.swift
//
//  XUPositionedWindowView.swift
//  XUCore
//
//  Created by Charlie Monroe on 11/5/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Cocoa

public class XUPositionedWindowView: NSView {

	@IBOutlet weak var connectedToWindow: NSWindow!

	public override func awakeFromNib() {
		if self.shouldBeConnectedOnAwakeFromNib {
			var superview = connectedToWindow.contentView?.superview
			if connectedToWindow.titlebarAccessoryViewControllers.count == 0 {
				/** Since 10.10.2, the titlebar view is deferred in creation.
				 * We need to force the creation by adding a blank titlebar accessory
				 * view controller.
				 */
				let controller = NSTitlebarAccessoryViewController()
				controller.layoutAttribute = .Right
				controller.view = NSView()
				connectedToWindow.addTitlebarAccessoryViewController(controller)
			}

			superview = superview?.subviews.find({ (view) -> Bool in
				return view.isKindOfClass(NSClassFromString("NSTitlebarContainerView")!)
			})
			superview = superview?.subviews.find({ (view) -> Bool in
				return view.isKindOfClass(NSClassFromString("NSTitlebarView")!)
			})

			superview?.addSubview(self)
		}

		NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(XUPositionedWindowView.updateFramePosition), name: NSWindowDidResizeNotification, object: connectedToWindow)

		self.postsFrameChangedNotifications = false

		self.updateFramePosition()
		self.localizeView()
	}

	deinit {
		NSNotificationCenter.defaultCenter().removeObserver(self)
	}

	public override var frame: CGRect {
		get {
			if self.superview == nil {
				return super.frame
			}
			return self.frameForWindowBounds(self.superview!.bounds, andRealFrame: super.frame)
		}
		set(newValue) {
			super.frame = newValue
		}
	}

	/*
	 func layoutAttribute() -> NSLayoutAttribute {
	 return NSLayoutAttributeBottom

	 }
	 */

	/// Override this to NO if you don't want the view to be connected to the
	/// window on awakeFromNib under some cirtumstances. By default YES.
	public var shouldBeConnectedOnAwakeFromNib: Bool {
		return true
	}

	/// Override this to specify position of the view in window
	public func frameForWindowBounds(bounds: CGRect, andRealFrame realFrame: CGRect) -> CGRect {
		XUThrowAbstractException()
	}
	
	/// Updates frame position. Shouldn't be invoked manually.
	@objc public func updateFramePosition() {
		let rect = self.frame
		self.frame = rect
		self.autoresizingMask = .ViewMinXMargin
	}
}

public class XUTopCenterWindowView: XUPositionedWindowView {

	public override func frameForWindowBounds(bounds: CGRect, andRealFrame realFrame: CGRect) -> CGRect {
		var retRect = CGRect(x: (bounds.width - realFrame.width) / 2.0, y: bounds.height - realFrame.height, width: realFrame.width, height: realFrame.height)
		retRect = retRect.integral
		return retRect
	}
}

public class XUTopLeftWindowCornerView: XUPositionedWindowView {

	public override func frameForWindowBounds(bounds: CGRect, andRealFrame realFrame: CGRect) -> CGRect {
		let retRect = CGRect(x: 0.0, y: bounds.height - realFrame.height, width: realFrame.width, height: realFrame.height)
		return retRect
	}
}

public class XUTopRightWindowCornerView: XUPositionedWindowView {

	public override func frameForWindowBounds(bounds: CGRect, andRealFrame realFrame: CGRect) -> CGRect {
		var retRect = CGRect(x: bounds.width - realFrame.width - 5.0, y: bounds.height - realFrame.height, width: realFrame.width, height: realFrame.height)

		retRect = retRect.integral

		retRect.origin.y -= 1.0

		return retRect
	}
}

